pkg/controller/snapshotter/snapshotter.go

```diff
@@ -273,38 +286,72 @@ func (vs *volumeSnapshotter) getSnapshotCreateFunc(snapshotName string, snapshot
 			Metadata: metav1.ObjectMeta{
 				Name: snapName,
 			},
-			Spec: crdv1.VolumeSnapshotDataSpec{
-				VolumeSnapshotRef: &v1.ObjectReference{
-					Kind: "VolumeSnapshot",
-					Name: snapshotName,
-				},
-				PersistentVolumeRef: &v1.ObjectReference{
-					Kind: "PersistentVolume",
-					Name: pvName,
-				},
-				VolumeSnapshotDataSource: *snapshotDataSource,
-			},
 			Status: crdv1.VolumeSnapshotDataStatus{
 				Conditions: []crdv1.VolumeSnapshotDataCondition{
 					readyCondition,
 				},
 			},
 		}
+
 		var result crdv1.VolumeSnapshotData
-		err = vs.restClient.Post().
-			Resource(crdv1.VolumeSnapshotDataResourcePlural).
-			Namespace(v1.NamespaceDefault).
-			Body(snapshotData).
-			Do().Into(&result)
+		for i := 0; i < createVolumeSnapshotDataRetryCount; i++ {
+			err = vs.restClient.Post().
+				Resource(crdv1.VolumeSnapshotDataResourcePlural).
+				Namespace(v1.NamespaceDefault).
+				Body(snapshotData).
+				Do().Into(&result)
+			if err != nil {
+				// Re-Try it as errors writing to the API server are common
+				glog.Infof("Error creating the VolumeSnapshotData %s: %v. Retrying...", snapshotName, err)
+				time.Sleep(createVolumeSnapshotDataInterval)
+			} else {
+				break
+			}
+		}
 
 		if err != nil {
-			//FIXME: Errors writing to the API server are common: this needs to be re-tried
 			glog.Errorf("Error creating the VolumeSnapshotData %s: %v", snapshotName, err)
+			// Don't proceed to create snapshot using the plugin due to error creating
+			// VolumeSnapshotData
+			return fmt.Errorf("Failed to create the VolumeSnapshotData %s for snapshot %s", snapName, snapshotName)
 		}
+
+		// Create a snapshot on the storage backend using the plugin
+		snapshotDataSource, err := vs.takeSnapshot(pv)
```

Comment:
A revised version:
- construct snapshot metadata info to help to uniquely identify a snapshot, such metadata can be the name of the `VolumeSnapshot` to fill into AWS snapshot description or a tag.
- if `VolumeSnapshot` binds to an existing `VolumeSnapshotData` that has the same metadata info with ready condition, then don't take more snapshot
otherwise:
- if `VolumeSnapshot` binds to an existing `VolumeSnapshotData` but the condition is not ready
- query backend to find if any snapshot has the same metadata
- if so,  create `VolumeSnapshotData` to point to this snapshot, update status in `VolumeSnapshotData`, update `VolumeSnapshot` to reference this `VolumeSnapshotData`
- post both `VolumeSnapshotData` and `VolumeSnapshot`

if no such snapshot is found, do the following:
- create `VolumeSnapshotData` (but don't post it yet)
- fill the metadata info in `VolumeSnapshotData`
- update `VolumeSnapshot` with `VolumeSnapshotData`, but don't post them yet
- take a snapshot and use this metadata info in the backend specfic snapshot object
- if snapshot is created successfully, update status in `VolumeSnapshotData`  and post `VolumeSnapshotData` and post the update to `VolumeSnapshot`

In this process, API server is posted 2 times. It is crash safe to me. Anything thought @tsmetana @jingxu97 @xing-yang ?